src/irmin-pack/inode.ml

```diff
@@ -1068,47 +1214,82 @@ struct
           (* This function is called when we encounter a Broken pointer with
              Truncated layouts. *)
-          if not @@ mem h then
-            Fmt.failwith
-              "You are trying to save to the backend an inode deserialized \
-               using [Irmin.Type] that used to contain pointer(s) to inodes \
-               which are unknown to the backend. Hash: %a"
-              pp_hash h
-          else
-            (* The backend already knows this target inode, there is no need to
-               traverse further down. This happens during the unit tests. *)
-            ()
+          match index h with
+          | None ->
+              Fmt.failwith
+                "You are trying to save to the backend an inode deserialized \
+                 using [Irmin.Type] that used to contain pointer(s) to inodes \
+                 which are unknown to the backend. Hash: %a"
+                pp_hash h
+          | Some key ->
+              (* The backend already knows this target inode, there is no need to
+                 traverse further down. This happens during the unit tests. *)
+              k key
         in
-        fun save_dirty arr ->
-          let iter_ptr = Ptr.save ~broken ~save_dirty ~clear layout in
+        fun ~save_dirty arr ->
+          let iter_ptr =
+            Ptr.save ~broken:{ f = broken } ~save_dirty ~clear layout
+          in
           Array.iter (Option.iter iter_ptr) arr
       in
       let rec aux ~depth t =
         [%log.debug "save depth:%d" depth];
         match t.v with
-        | Values _ -> add (Lazy.force t.hash) (to_bin layout t)
+        | Values _ -> (
+            let unguarded_add hash =
+              let value =
+                (* NOTE: the choice of [Bin.mode] is irrelevant (and this
+                   conversion is always safe), since nodes of kind [Values _]
+                   contain no internal pointers. *)
+                to_bin layout Bin.Ptr_key t
+              in
+              let key = add hash value in
+              Val_ref.promote_exn t.v_ref key;
+              key
+            in
+            match Val_ref.inspect t.v_ref with
+            | Key key ->
+                if mem key then key else unguarded_add (Key.to_hash key)
+            | Hash hash -> unguarded_add (Lazy.force hash))
         | Tree n ->
-            iter_entries
-              (fun t ->
-                let hash = Lazy.force t.hash in
-                if mem hash then () else aux ~depth:(depth + 1) t)
-              n.entries;
-            add (Lazy.force t.hash) (to_bin layout t)
+            let save_dirty t k =
+              let key =
+                match Val_ref.inspect t.v_ref with
+                | Key key -> if mem key then key else aux ~depth:(depth + 1) t
+                | Hash hash -> (
+                    match index (Lazy.force hash) with
+                    | Some key ->
+                        if mem key then key else aux ~depth:(depth + 1) t
```

Comment:
This branch seem to be reached only during the `PACK MEM` UTs

Reply:
Indeed, this follows the general contract of [`index`](https://github.com/mirage/irmin/blob/eb12b83ee62ee21fb7bf2434c3cd31c602dc9205/src/irmin/indexable_intf.ml#L42-L46) functions on Irmin backend stores rather than the specific one for `Pack_store.index`, so we do some extra work (or extra integrity-checking) here in the standard `irmin-pack` deployment. We could avoid that by either (a) adding an extra function to `Indexable.S` that chains `index` and `mem`, or (b) abstracting over this difference in semantics so that `PACK MEM` can do what it needs to do.

Have added a comment for now.